text/0000-cloud-based-mapping.md

```diff
@@ -0,0 +1,134 @@
+
+This will require us to
+
+- extract interfaces out of `MapBuilderBridge` and `SensorBridge`.
+- implement a gRPC bridging version of the above two classes that serialize any sensor data to proto3 and send that to a gRPC service.
+
+In principle once we have a binary that does ROS to gRPC translation it would be possible to replace the existing cartographer ROS node with a combination of this translation binary and *cartographer_grpc*, i.e. instead of the C++ interface between *cartographer_node* and libcartographer that exists today we would introduce a local, loopback gRPC layer.
+We have decided not to go this route and keep the current *cartographer_node* as is until we have a better idea of the performance impact (or benefit) of this additional gRPC serialization / deserialization and whether adding gRPC to our dependencies is acceptable for our ROS users.
+
+#### cartographer_grpc
+
+We introduce a new binary,  *cartographer_grpc*, in the *cartographer* repo, which offers the cartographer gRPC service - essentially a (flattened) gRPC representation of the `MapBuilder` interface. 
+This binary runs both on agent and cloud albeit with different runtime configuration.
+Depending on its configuration it acts both as gRPC server and client. It can be configured to
+
+- run as a standalone SLAM solution that receives sensor data as gRPC messages and publishes SLAM results as gRPC messages.
+This is essentially a thin wrapper around the C++ API of libcartographer.
+- run as a purely localizing instance that receives sensor data as gRPC messages, publishes the robot pose as gRPC messages and upstreams local SLAM results to the cloud as gRPC messages.
+- offers a gRPC service that accepts local SLAM results and only solves the global SLAM problem, offering the SLAM solution as gRPC messages.
+
+Upon installation *cartographer_grpc* installs the generated gRPC service headers for e.g. binaries in cartographer_ros.
+
+| name                             | repo             | dependencies      |
+| -------------------------------- | ---------------- | ----------------- | 
+| libcartographer                  | cartographer     | proto3            |
+| cartographer_grpc                | cartographer     | proto3, gRPC      |
+| cartographer_(offline)_node      | cartographer_ros | proto3, ROS       |
+| cartographer_(offline)_node_grpc | cartographer_ros | proto3, gRPC, ROS |
+
+### Detailed Design
+
+#### cartographer_grpc
+
+The important interaction between local and global SLAM happens through `GlobalTrajectoryBuilder` between `LocalTrajectoryBuilder` and `PoseGraph`. 
+
+The following component diagram describes the cloud mapping architecture
```

Comment:
I guess the right side is "Cloud Cartographer" while the left one is "Agent Cartographer". Might make sense to point that out.

Reply:
Thanks. Done.